### ISSUE TITLE
[Feature Request] Support for Ray backend and Ray Tune

### ISSUE THREAD

**Issue by rohan-gt** [pycaret/pycaret#428]
Are there any plans for supporting something like a [Ray](https://github.com/ray-project/ray) backend for scaling up the ML in general and Ray Tune using [tune_sklearn](https://github.com/ray-project/tune-sklearn) which can scale up hyperparameter tuning and also support Bayesian optimization for tuning?

**Comment by pycaret** [comment]
@rohan-gt Would you like to work on this?

**Comment by Yard1** [comment]
If @rohan-gt isn't interested in working in it we could add it during the refactor, @pycaret

**Comment by pycaret** [comment]
@Yard1 Sounds good.

**Comment by Yard1** [comment]
I took a look at Tune (with tune_sklearn). It uses scikit-optimize underneath, which doesn't support the latest scikit-learn. I think we may have to think about, at least temporarily, writing our own bayes optimization function.

**Comment by pycaret** [comment]
@Yard1 Yes. I don't think `scikit-optimize` is an active project. We will be better off writing our own Bayes optimizer if we can.

**Comment by rohan-gt** [comment]
@pycaret sorry for the late reply. I won't be able to work on this but as a suggestion using the [TuneBOHB](https://docs.ray.io/en/master/tune/api_docs/suggestion.html#bohb-tune-suggest-bohb-tunebohb) algorithm along with the [HyperBandForBOHB](https://docs.ray.io/en/master/tune/api_docs/schedulers.html#bohb-tune-schedulers-hyperbandforbohb) scheduler seems to be the best combination

**Comment by Yard1** [comment]
It looks like we could fork tune-sklearn and only replace the optimizer used with another one (like TuneBOHB) from ray.suggest. Since they all use the same API, it should be just a matter of converting the param grid to the specified format and other minor tweaks.

**Comment by pycaret** [comment]
@Yard1 This sounds good to me.

**Comment by Yard1** [comment]
I have toyed around with it and I believe I have managed to edit tune-sklearn to use TuneBOHB, making it compatible with the latest scikit-learn and thus pycaret. It was not hard, most of the heavy lifting was already done. I will now try to implement it in pycaret's `tune_model`s - if all goes well, we can drop it in with 2.2.

Would love to discuss how we should distribute this, though @pycaret - we can either make a separate package based on tune-sklearn and upload it to pip, or make tune-sklearn an (optional) dependency for pycaret, and overwrite the classes in it inside pycaret.

**Comment by pycaret** [comment]
@Yard1 Yes. Let's drop this in `2.2`

**Comment by Yard1** [comment]
The best option would be for `tune-sklearn` team to integrate it. I'll open up a PR when I am done. I don't know how often they release, but it just needs to happen before PyCaret 2.2, which isn't coming soon anyway.

**Comment by Yard1** [comment]
Made a PR to `tune-sklearn` - https://github.com/ray-project/tune-sklearn/pull/68 - hopefully this gets accepted so we can use it without having to worry about maintaining our own fork.

**Comment by holgern** [comment]
@Yard1  The newest scikit-optimize version 0.8.1 supports scikit-learn >=23, the new version was released today.

**Comment by Yard1** [comment]
Thanks, we'll add that as an option :+1:

**Comment by pycaret** [comment]
@rohan-gt 

This issue has been fixed in the `pycaret-nightly` release. Please install pycaret-nightly using `pip install pycaret-nightly`. Feel free to open a new issue, in case the problem persists. This fix will be published to PyPi as `pycaret==2.2.0` on October 28, 2020.
